### ISSUE TITLE
Data loss: Image files added to taxonomy term descriptions do not get a usage count and are then deleted by garbage collection

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5016]
This issue here is related to #5007, and the use case below is about images added to the "Description" field of taxonomy terms or the "Block content" field in custom blocks, but it is broader than that.

**Steps To Reproduce**

To reproduce the behavior:
1. In a vanilla installation of Backdrop, go to `admin/content/files` and use the "+ Add file" link to upload an image.
2. After uploading the image successfully, you are redirected to the file listing page, where you notice that the file status is set to "Permanent", and its usage count is 1 👍🏼 
3. Start creating a post node (`node/add/post`)
4. Add a title.
5. Click the "Insert image" button in the WYSIWYG.
6. Switch to the "Select from library" tab.
7. Select the image file you uploaded in step 1 -> click "insert"
8. Save the node.
9. Head to the file listing page (`admin/content/files`) -> notice that the image usage count has now increased to 2 👍🏼 
10. Edit the post node you created in steps 3 to 8.
11. Click the "Select existing file" button in the "Image" field.
12. Select the image file you uploaded in step 1 -> click "insert"
13. Save the node.
14. Head to the file listing page -> notice that the image usage count has now increased to 3 👍🏼 
15. View the post node you created, and scroll to the bottom of the page to add a comment.
16. In the comment field, click the "Insert image" button in the WYSIWYG.
17. Switch to the "Select from library" tab.
18. Select the image file you uploaded in step 1 -> click "insert"
19. Save the comment.
20. Head to the file listing page -> notice that the image usage count has now increased to 4 👍🏼 
21. Go to `admin/structure/taxonomy/tags/add`, to add a taxonomy term in the "Tags" vocabulary.
22. Add a title for the term.
23. In the "Description" field of the term, click the "Insert image" button in the WYSIWYG (there is no "Upload" tab in the dialog, but that's a separate issue: #5007 - ignore this)
24. Select the image file you uploaded in step 1 -> click "insert"
25. Save the term.
26. Head to the file listing page -> notice that the image usage count is still 4 👎🏼 
27. Edit the default layout, and add a custom block to the sidebar region.
28. Click the "Insert image" button in the WYSIWYG of the "Block content" field.
29. Switch to the "Select from library" tab.
30. Select the image file you uploaded in step 1 -> click "insert"
31. Save the block (you may leave the display title of the block blank)
32. Save the layout.
33. Head to the file listing page -> notice that the image usage count is still 4 👎🏼 

**Actual behavior**

Image files added to the taxonomy term "Description" field or the "Block content" field in custom blocks do not get a usage count. If you add existing images, their usage count doesn't get incremented.

**Expected behavior**

Image files added to these fields should get their usage count added/incremented.

---

** Additional WTF!! 😅 **

Go back and:

1. Delete the comment you added to the post you created.
2. Head to the file listing page -> notice that the image usage count has been decreased to 3 👍🏼 
3. Edit the post node, and remove the image from the "Image" field -> save the node.
4. Head to the file listing page -> notice that the image usage count has been decreased to 2 👍🏼 
5. Edit the post node again, and remove the image from the "Body" field -> save the node.
6. Head to the file listing page -> notice that the image usage count has been decreased to 1 👍🏼 

At this point, novice users that do not know exactly how files and media/file entities work in Drupal/Backdrop, will wonder where that single file usage is coming from. Expected WTF?! (separate issue)

Experienced Backdrop/Drupal users will think that this usage count is coming from the file entity, so they might head there and delete the file.

1. Head to `admin/content/files` and use the "Delete" action on the image file.
2. After deleting the image successfully, you are redirected to the file listing page, where you are told that there are "No files available." ...at this point, you would expect that the site would have ended up with a broken term description, and a broken block. Right?
3. Head to the term listing for the "Tags" vocabulary (`admin/structure/taxonomy/tags`).
4. Click the link to view the term you have added earlier -> both the image in the term description, as well as in the custom block you added in the sidebar are there!!:

<img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/111876760-ea19c300-89a8-11eb-9d06-392d39ee5a13.png">

**Comment by indigoxela** [comment]
@klonos let's try to drill this down a bit.

Re "left over" usage count when a file has been uploaded via "add file" link: That's correct behavior IMO. Ironically the PR was from you. :wink: https://github.com/backdrop/backdrop-issues/issues/4017
The "File" is self-referencing, otherwise it couldn't be used at all.

Re images inserted into custom blocks: This might need some more testing. There's an older issue that claims to have fixed that.
https://github.com/backdrop/backdrop-issues/issues/1373

Re image usage recording in taxonomy term description pseudo fields: that's a real problem and probably the main thing we should fix here. Currently the taxonomy module does not bother at all, but it should as it provides the field.

**Comment by klonos** [comment]
> Re images inserted into custom blocks: This might need some more testing.

You can confirm this easily:
1. Upload an file to the site via "+Add file" (`file/add`)
2. Confirm that the file is added to `admin/content/files` as permanent, with usage count 1 👍🏼 
3. Create a custom block, either from a layout, or from `admin/structure/block/add`.
4. Add the image you've uploaded in step 1 to the "Block content" field.
5. Go to `admin/content/files` and see that the image file count is still 1 👎🏼 
6. Delete the managed file, and confirm that there are no files listed in `admin/content/files`.
7. Go to `admin/structure/block` and edit the custom block -> the image is still there in the "Block content field"

So you basically have images added to the custom block content field, that are not added to the managed files, and do not get their usage count incremented/decreased as they are being added/removed to custom blocks.

This is the exact same behavior as with the "Description" field in taxonomy terms.

**Comment by indigoxela** [comment]
Re blocks: I wonder, why this [file_usage_add](https://github.com/backdrop/backdrop/blob/72b071a1842d9b745fc0fa28c069b20c6260984b/core/modules/block/block.module#L424) fails to do the job in this case.

**Comment by klonos** [comment]
> although both modules show the same (or very similar) behavior, I'm not sure yet, if we should cover both of them (Layout and Taxonomy) in the same issue.

Agreed, if this turns out to be a completely different fix for these 2 cases, then we should split this. If it is either a fix implemented in a single place, or the same fix applied to two places, then I think that we should fix it in a single PR, in one go.

**Comment by klonos** [comment]
We discussed this issue with @hosef during office hours today, and it is broader than the specific use cases that I have initially raised this for. So I've updated the issue title accordingly.

**Comment by indigoxela** [comment]
> it is broader than the specific use cases that I have initially raised this for.

Not sure if it's a good idea to shift the purpose/topic. IMO the problem with the taxonomy term description and the problem with config (like custom reusable blocks) differ quite a bit and need different solutions.

For taxonomy terms - which *are* fieldable entities - the solution is to get rid of that old property, which predates to inclusion of CCK fields in core and also predates the conversion of terms to fully classed entities. And instead use a proper field. In a backwards compatible way (with a migration step).

For config items the problem is trickier, as they lack the hooks for comparing previous/current versions. That makes *updating* the usage value accordingly quite difficult.

**Comment by klonos** [comment]
Thanks @indigoxela ...for completeness, copy/pasting some comments from the [recent Zulip discussion about this](https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/CKEditor.205/near/445474110):

@hosef:
> I think that supporting entity properties is very important for larger sites. It is important for reducing query load by putting things into the main entity table.
>
> It is definitely not a relic, and I think it's even more common in D9+ than it was in D6 and D7

**Comment by indigoxela** [comment]
@klonos you managed to paste one person's opinion re that pseudo field specific to taxonomy terms, but out of context I find it rather misleading. Even more, as the topic here shifted recently.

**Comment by hosef** [comment]
We talked a little bit about this at the last meeting. It was a bit all over the place and it seems like there are a bunch of different issues wrapped up in this, so I am going to try and summarize them.

**Changing Term description from an entity property to a field**
It seemed like most people thought this was a good idea, but we probably can't do this in 1.x because removing properties that are available on a class would be an API breaking change.

**Providing a generic way of handling entity properties that use a rich text editor**
We can't do this currently, because we don't have enough information about entity properties in core. The entity property API was not added to core, so anything that needs to know about the properties would have to depend on Entity Plus.

**Making the Taxonomy module handle this for taxonomy terms**
This is what Nate thought would be the most realistic option for us to fix the issue in the short term. I have provided a PR that fixes the issue.

**Handling cases where config values are edited with a rich text editor**
This is going to be a pretty complicated because files referenced in config are not synchronized with the config. You can end up with states where files added by config values may or may not exist on any other environment and files with the same id might not actually be the same file on other environments.
Probably the best option would be to just make all files uploaded in these cases permanent. This would stop Backdrop from deleting files that are in use, but it would also require manual intervention from users to clean up files that are no longer needed.

**Comment by argiepiano** [comment]
> Handling cases where config values are edited with a rich text editor...
Probably the best option would be to just make all files uploaded in these cases permanent. 

@hosef I may not be understanding this problem. If I upload an image in a Custom block, that image is currently made Permanent. Is that not what you are seeing? For example, if I go to Layout, choose to modify the Default layout, add a custom block and upload an image, then go to `admin/content/files`, that uploaded file is Permanent. It's only Temporary if I close and not save the Layout. Please clarify.

**Comment by argiepiano** [comment]
> Making the Taxonomy module handle this for taxonomy terms
This is what Nate thought would be the most realistic option for us to fix the issue in the short term. 

Agreed. I have reviewed and tested PR number 5106. LGTM and RTBC.

Can I suggest we split this issue, since there are different aspects to fix? The PR fixes the Taxonomy Term description question only.


Side note:
>Providing a generic way of handling entity properties that use a rich text editor
We can't do this currently, because we don't have enough information about entity properties in core. The entity property API was not added to core, so anything that needs to know about the properties would have to depend on Entity Plus.

Agreed too. I've always lamented the loss of several aspects of Entity API in Backdrop. There was a level of abstraction that was lost. Something to consider for 2.0.

**Comment by quicksketch** [comment]
Code-wise this looks good to solve the Taxonomy Term Description + File Usage. However we need test coverage on something like this that is so easily overlooked.

**Comment by hosef** [comment]
@argiepiano I just tested and it did make them permanent, so something must have changed since the issue was initially created. I have been mostly focused on the Taxonomy issue since that is the one that is affecting me at the moment.

**Comment by klonos** [comment]
> Can I suggest we split this issue, since there are different aspects to fix? The PR fixes the Taxonomy Term description question only.

@argiepiano yes 👍🏼 ...that is what @quicksketch suggested during the dev meeting as well.

This issue was originally created with two use cases:
- Taxonomy term descriptions: Should be handled here.
- Block descriptions: Should be separate issue.

...then the use cases of files in config and entity properties should also be dealt with separately.

I'll loop back to this and clean things up properly. For now, I'll only adjust the issue title here.

**Comment by quicksketch** [comment]
This code is good but still needs test coverage, with @hosef and @klonos on breaks, I'll try to see if I can pick this up and complete the test coverage.
